keps/sig-node/4383-dra-semantic-model-mode-switching/README.md

```diff
@@ -0,0 +1,898 @@
+
+Each instance can be only in one mode at once. It can switch between modes only
+when it is not currently in use.
+
+### Risks and Mitigations
+
+Adding another semantic model doesn't add any additional risks. There is a
+slight performance impact when checking whether the builtin controller of the
+semantic model needs to do something for a claim or NodeResourceCapacity object,
+but that impact should be minimal when the model is not in use.
+
+This resource model is fairly complex. Finding the right instance and mode of
+that instance to satisfy a request has to be done by iterating over them, which
+can be expensive. To mitigate that, requests should be written so that "simple"
+checks are executed first to prune the search tree early.
+
+## Design Details
+
+The semantic model is hosted in
+`k/k/staging/src/k8s.io/dynamic-resource-allocation`:
+
+- `semantic/modeswitch`: the implementation of the builtin controller
+- `apis/modeswitch`: the versioned type(s)
+- `internal/apis/modeswitch`: the internal type for conversion and for use
+  by the implementation
+
+The only modification elsewhere is registering the builtin controller in the
+dynamic resource scheduler plugin if (and only if) the model is enabled.
+
+The semantic model uses a label selector for selecting instances and modes that
+is more capable than `metav1.LabelSelector`. For example, semantic version
+comparison and nested expressions are supported. This label selector is defined
+in its own packages and can be reused by future semantic models, if desired:
+
+- `semantic/labels`: the implementation of the label selector
+- `apis/labels`: the versioned type(s)
+- `internal/apis/labels`: the internal type for conversion and for use
+  by the implementation
+
+### Label selector
+
+The label selector is matched against a set of label name and value pairs. In
+the mode switching model, this set is the combination of labels specified for
+the instance and the labels specified for each mode, with labels of a mode
+overwriting the ones of the instance.
+
+The selector supports expressions like `"capabilityFoo" || (!"capabilityBar" && "name" >= "a" && "memory" quantity< "20Gi" && "driver" version> "v1.0.0")`.
+
+The corresponding YAML representation of this example is:
+```
+selector:
+  label: capabilityFoo
+  op: IsSet
+  or:
+    - and:
+      - label: capabilityBar
+        op: IsNotSet
+      - label: name
+        op: GE
+        value: a
+      - label: memory
+        op: LT
+        value: 20Gi
+        type: Quantity
+      - label: driver
+        op: GT
+        value: v1.0.0
+        type: Semver
+```
+
+The underlying Go type is:
+```
+type Selector struct {
+    Label string
+    Op    SelectorOp // LT, LEQ, EQ, NEQ, GEQ, GT, IsSet, IsNotSet
+    Type  ValueType  // String (default), Quantity (= resource.Quantity), Semver (semantic version number)
+    Value string
+    And   []Selector
+    Or    []Selector
+}
```

Comment:
@thockin: in https://github.com/kubernetes/enhancements/pull/4384/files#r1473463075 you wrote
>  Encoding an expression-grammar in YAML is kind of disgusting.

So `metav1.LabelSelector` is disgusting? It does exactly that. This `Selector` is conceptually very similar, just with additional operations, type support, and nested expression. Okay, quite a bit more complex... But remember that this is for DRA driver vendors, not end-users.

I'm open for suggestions if there is a better way to do this while preserving versioning support in the API.

Reply:
> So metav1.LabelSelector is disgusting?

Yes, it is :)  It's mitigated by the limited scope (no compound booleans, limited operators), but I feel that we should have just made it an expression (we didn't have CEL at the time, but still).

```
selector: In(self["example.com/my-key"], ["foo", "bar"])
```

vs:

```
selector:
  matchExpressions:                                                                                        
    - key: "example.com/my-key"
      operator:  NotIn
      values:
      - foo
      - bar
 ```

Reply:
Ack. I'm sold on the idea of using CEL expressions instead.